pipeline: filters: lookup added new filter page (#1953)
fluent/fluent-bit-docs: pipeline/filters/lookup.md

# Lookup

> The Lookup plugin looks up a key value from a record in a specified CSV file and, if a match is found, adds the corresponding value from the CSV as a new key-value pair to the record.

Author: This might be a little confusing - it does not use the key from CSV file - it uses the result_key input as the key and the CSV value (i.e. second column) as the value and adds that combination to the record.

Thank you for all the other clarifications.

> The Lookup plugin looks up a key value from a record in a specified CSV file and, if a match is found, adds the corresponding value from the CSV as a new key-value pair to the record.

Reviewer: ah, that makes sense! I'll clarify some of those suggestions accordingly...

## Configuration parameters

The plugin supports the following configuration parameters

| Key | Description | Default |
| :-- | :---------- | :------ |
| `file` | The CSV file that Fluent Bit will use as a lookup table. The file should contain two columns (key and value), with the first row as an optional header that is skipped. Supports quoted fields and escaped quotes. | _none_ |
| `lookup_key` | The specific key in the input record to look up in the CSV file's first column. Supports [record accessor](../../administration/configuring-fluent-bit/record-accessor). | _none_ |
| `result_key` | The name of the key to add to the output record with the matched value from the CSV file's second column if a match is found. | _none_ |

> | `ignore_case` | Ignore case when matching the lookup key against the CSV keys. | `false` |

Reviewer: ```suggestion
| `ignore_case` | Specifies whether to ignore case when searching for `lookup_key`. If `true`, searches are case-insensitive. If `false`, searches are case-sensitive. Possible values: `true`, `false`. | `false` |
```

Author: The ignore_case is only applicable when matching the lookup_key value to the first column of the CSV (which is the key for the CSV). I think the change here can be interpreted that ignore_case is for how the lookup_key is matched against record rather than CSV. It's quite a tricky thing to describe actually, but flow is lookup_key defines a record assessor that will be used to find a value in the record; that value is then matched against the first column of the CSV; if a match is found the corresponding value for the matched row is taken from the second column; this extracted value from the second column is then added to the record under a key defined under result_key.

Reviewer: good callout... let me chew on that for a minute and see if I can suggest something :P

## Example configuration

{% tabs %}
{% tab title="fluent-bit.yaml" %}

```yaml
parsers:
  - name: json
    format: json

pipeline:
  inputs:
    - name: tail
      tag: test
      path: devices.log
      read_from_head: true
      parser: json

  filters:
    - name: lookup
      match: test
      file: device-bu.csv
      lookup_key: $hostname
      result_key: business_line
      ignore_case: true

  outputs:
    - name: stdout
      match: test
```

{% endtab %}
{% tab title="fluent-bit.conf" %}

```text
[PARSER]
    Name        json
    Format      json

[INPUT]
    Name              tail
    Tag               test
    Path              devices.log
    Read_from_head    On
    Parser            json

[FILTER]
    Name           lookup
    Match          test
    File           device-bu.csv
    Lookup_key     $hostname
    Result_key     business_line
    Ignore_case    On

[OUTPUT]
    Name   stdout
    Match  test
```

{% endtab %}
{% endtabs %}

> The following configuration reads log records from `devices.log` that includes the following values for device hostnames:

Reviewer: ```suggestion
The previous configuration reads log records from `devices.log` that includes the following values for device hostnames:
```

```text
{"hostname": "server-prod-001"}
{"hostname": "Server-Prod-001"}
{"hostname": "db-test-abc"}
{"hostname": 123}
{"hostname": true}
{"hostname": " host with space "}
{"hostname": "quoted \"host\""}
{"hostname": "unknown-host"}
{}
{"hostname": [1,2,3]}
{"hostname": {"sub": "val"}}
{"hostname": " "}
```

> It uses the value of the `hostname` field (which has been set as the `lookup_key`) to find matching values in column 1 of the  (`device-bu.csv`) CSV file.

Reviewer: ```suggestion
It uses the value of the `hostname` field (which was set as the `lookup_key`) to find matching values in the first column of `device-bu.csv`.
```

```text
hostname,business_line
server-prod-001,Finance
db-test-abc,Engineering
db-test-abc,Marketing
web-frontend-xyz,Marketing
app-backend-123,Operations
"legacy-system true","Legacy IT"
" host with space ","Infrastructure"
"quoted ""host""", "R&D"
no-match-host,Should Not Appear
```

> Where a match is found the filter adds new key (name of which is set by the `result_key` input) with the value from the second column of the CSV file of the matched row.

Reviewer: ```suggestion
When the filter finds a match, it adds a new key with the name specified by `result_key` and a value from the second column of the CSV file of the row where `lookup_key` was found.
```

> For above configuration the following output can be expected (when matching case is ignored as `ignore_case` is set to true):

Reviewer: ```suggestion
This results in output that resembles the following:
```

# Lookup

> The Lookup plugin looks up a key value from a record in a specified CSV file and, if a match is found, adds the corresponding value from the CSV as a new key-value pair to the record.

Reviewer: ```suggestion
The _Lookup_ plugin searches a CSV file for values that match the value of a specific key in a record and then, if it finds a match, adds a new key to the record whose value matches the corresponding CSV value.
```

## Configuration parameters

The plugin supports the following configuration parameters:

| Key | Description | Default |
| :-- | :---------- | :------ |

> | `file` | The CSV file that the Lookup filter will use as a lookup table. This file must contain one column of keys and one column of values. The Lookup filter treats the first row as a header and ignores it. Supported quoted fields and escaped quotes. | _none_ |

Reviewer: ```suggestion
| `file` | The CSV file that the Lookup filter will use as a lookup table. This file must contain one column of keys and one column of values. The Lookup filter treats the first row as a header and ignores these values. Supports quoted fields and escaped quotes. | _none_ |
```

> | `lookup_key` | Specifies the key to search for in the CSV file's first column. Supports [record accessor](../administration/configuring-fluent-bit/classic-mode/record-accessor) syntax. | _none_ |

Reviewer: ```suggestion
| `lookup_key` | Specifies the record key whose value search for in the CSV file's first column. Supports [record accessor](../administration/configuring-fluent-bit/classic-mode/record-accessor) syntax. | _none_ |
```

> | `result_key` | If `lookup_key` is found, sets the name of the new key to add to the output record. This new key uses the corresponding value from the second column of the CSV file in the same row where `lookup_key` was found.  | _none_ |

Reviewer: ```suggestion
| `result_key` | If `lookup_key` is found, specifies the name of the new key to add to the output record. This new key uses the corresponding value from the second column of the CSV file in the same row where `lookup_key` was found.  | _none_ |
```
(slightly tweaking my own suggestion, lol)

> | `result_key` | If `lookup_key` is found, specifies the name of the new key to add to the output record. This new key uses the corresponding value from the second column of the CSV file in the same row where `lookup_key` was found.  | _none_ |

Reviewer: ```suggestion
| `result_key` | If a CSV entry whose value matches the value of `lookup_key` is found, specifies the name of the new key to add to the output record. This new key uses the corresponding value from the second column of the CSV file in the same row where `lookup_key` was found.  | _none_ |
```

> | `ignore_case` | Ignore case when matching the lookup key against the CSV keys. | `false` |

Reviewer: ```suggestion
| `ignore_case` | Specifies whether to ignore case when searching for the value of `lookup_key` in the specified CSV file. If `true`, searches are case-insensitive. If `false`, searches are case-sensitive. Possible values: `true`, `false`. | `false` |
```
apologies for the delay... so I think I may have fundamentally misunderstood how this filter works, which means I should update my other suggestions as well, lol. it this specific suggestion more accurate?

## Example configuration

{% tabs %}
{% tab title="fluent-bit.yaml" %}

```yaml
parsers:
  - name: json
    format: json

pipeline:
  inputs:
    - name: tail
      tag: test
      path: devices.log
      read_from_head: true
      parser: json

  filters:
    - name: lookup
      match: test
      file: device-bu.csv
      lookup_key: $hostname
      result_key: business_line
      ignore_case: true

  outputs:
    - name: stdout
      match: test
```

{% endtab %}
{% tab title="fluent-bit.conf" %}

```text
[PARSER]
    Name        json
    Format      json

[INPUT]
    Name              tail
    Tag               test
    Path              devices.log
    Read_from_head    On
    Parser            json

[FILTER]
    Name           lookup
    Match          test
    File           device-bu.csv
    Lookup_key     $hostname
    Result_key     business_line
    Ignore_case    On

[OUTPUT]
    Name   stdout
    Match  test
```

{% endtab %}
{% endtabs %}

> The previous configuration reads log records from `devices.log` that includes the following values for device hostnames:

Reviewer: ```suggestion
The previous configuration reads log records from `devices.log` that includes the following values in the `hostname` field:
```

> ```text

Reviewer: ```suggestion
```shell
```

{"hostname": "server-prod-001"}
{"hostname": "Server-Prod-001"}
{"hostname": "db-test-abc"}
{"hostname": 123}
{"hostname": true}
{"hostname": " host with space "}
{"hostname": "quoted \"host\""}
{"hostname": "unknown-host"}
{}
{"hostname": [1,2,3]}
{"hostname": {"sub": "val"}}
{"hostname": " "}
```

> It uses the value of the `hostname` field (which was set as the `lookup_key`) to find matching values in the first column of `device-bu.csv`.

Reviewer: ```suggestion
Because `hostname` was set as the `lookup_key`, the Lookup filter uses the value of each `hostname` key within the record to search for matching values in the first column of the CSV file.
```

```text

> hostname,business_line

Reviewer: ```suggestion
host,department
```
minor suggestion, but I wonder if it would make the examples clearer to use values for the CSV header row that are NOT found anywhere else. that way it doesn't seem like `hostname` or `business_line` are referring to the CSV header row items
